lib/common/ipc_server.c

```diff
@@ -620,45 +620,17 @@ pcmk__ipc_prepare_iov(uint32_t request, const xmlNode *message,
     header->size_uncompressed = 1 + buffer->len;
     total = iov[0].iov_len + header->size_uncompressed;
 
-    if (total < max_send_size) {
-        iov[1].iov_base = pcmk__str_copy(buffer->str);
-        iov[1].iov_len = header->size_uncompressed;
-
-    } else {
-        static unsigned int biggest = 0;
-
-        char *compressed = NULL;
-        unsigned int new_size = 0;
-
-        if (pcmk__compress(buffer->str,
-                           (unsigned int) header->size_uncompressed,
-                           (unsigned int) max_send_size, &compressed,
-                           &new_size) == pcmk_rc_ok) {
-
-            pcmk__set_ipc_flags(header->flags, "send data", crm_ipc_compressed);
-            header->size_compressed = new_size;
-
-            iov[1].iov_len = header->size_compressed;
-            iov[1].iov_base = compressed;
-
-            biggest = QB_MAX(header->size_compressed, biggest);
-
-        } else {
-            crm_log_xml_trace(message, "EMSGSIZE");
-            biggest = QB_MAX(header->size_uncompressed, biggest);
-
-            crm_err("Could not compress %u-byte message into less than IPC "
-                    "limit of %u bytes; set PCMK_ipc_buffer to higher value "
-                    "(%u bytes suggested)",
-                    header->size_uncompressed, max_send_size, 4 * biggest);
-
-            free(compressed);
-            pcmk_free_ipc_event(iov);
-            rc = EMSGSIZE;
-            goto done;
-        }
+    if (total >= max_send_size) {
+        crm_log_xml_trace(message, "EMSGSIZE");
+        crm_err("Could not transmit message; message size %u bytes is larger "
+                "than the maximum of %u", header->size_uncompressed, max_send_size);
+        rc = EMSGSIZE;
```

Comment:
`pcmk_free_ipc_event(iov)`; or move that to an `if (rc != pcmk_rc_ok)` block after the `done` label

Reply:
We are pretty inconsistent about this - for instance, `crm_ipc_send` doesn't free the iov at all on error, while `cib_notify_send` always frees it.  I'll add a patch on top that makes `pcmk__ipc_prepare_iov` free the vector on error and change the callers to respect that.